File: ShapeByte/shared/progress/data/Progress.swift
//
//  Progress.swift
//  ShapeByte
//
//  Created by Lang, Stefan [RTL Tech] on 28.07.24.
//

import Foundation

// TODO: test
struct Progress: Equatable {
    static let zero = Progress(0)

    let value: CGFloat // 0..1

    // TODO: consider method with rounding rule
    var absoluteValue: Int { // 0 .. 100
        Int((value * 100.0).rounded())
    }

    init(_ value: CGFloat) {
        self.value = Self.adjustProgress(value)
    }

    // TODO: test
    // TODO: consider rounding rule
    init(absoluteValue: Int) {
        let relativeValue = CGFloat(absoluteValue) / 100.0
        self.init(relativeValue)
    }

    private static func adjustProgress(_ progress: CGFloat) -> CGFloat {
        let retVal: CGFloat
        if progress > 1.0 {
            retVal = 1.0
        } else if progress < 0.0 {
            retVal = 0.0
        } else {
            retVal = progress
        }

        return retVal
    }
}

File: shared/core/utils/api/src/commonMain/kotlin/de/stefan/lang/coreutils/api/progress/Progress.kt
package de.stefan.lang.coreutils.api.progress
import kotlin.math.roundToInt
import kotlin.time.Duration

data class Progress(private val progress: Float) {
    companion object {
        val ZERO = Progress(0f)
        val COMPLETE = Progress(1f)

        const val ABSOLUTE: Int = 100

        // TODO: consider rounding rule
        fun with(current: Int, total: Int): Progress {
            val progress = current.toFloat() / total.toFloat()
            return Progress(progress)
        }

        // TODO: consider rounding rule
        // TODO: Test
        fun with(current: Float, total: Float): Progress {
            val progress = current / total
            return Progress(progress)
        }

        // TODO: consider rounding rule
        fun with(current: UInt, total: UInt): Progress {
            val progress = current.toFloat() / total.toFloat()
            return Progress(progress)
        }

        // TODO: consider rounding rule
        // TODO: Test
        fun with(current: Duration, total: Duration): Progress {
            val progress = current.inWholeMilliseconds.toFloat() /
                total.inWholeMilliseconds.toFloat()

            return Progress(progress)
        }
    }

    val value: Float = adjustProgress(progress)
    val absoluteValue: Int by lazy { (value * ABSOLUTE).roundToInt() }

    constructor(value: Int) : this(value.toFloat() / ABSOLUTE)

    // TODO: Test
    operator fun plus(progress: Progress): Progress = Progress(this.value + progress.value)

    private fun adjustProgress(value: Float): Float {
        return when {
            value < 0 -> 0f
            value > 1 -> 1f
            else -> value
        }
    }
}
